### ISSUE TITLE
Using XLA with subprocesses.Popen instead of spawn

### ISSUE THREAD

**Issue by tchaton** [pytorch/xla#2836]
## ❓ Questions and Help

Dear people from torch-xla,

PyTorch Lightning currently implements a TPUSpawn training_type_plugin relying on `xmp.spawn(self.new_process, **self.xmp_spawn_kwargs)`
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/training_type/tpu_spawn.py#L279 to launch training.

I would like to create a TPUPopen version relying on subprocessses.Popen, similar to PyTorch Lightning DDP vs DDPSpawn versions: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/training_type/ddp.py#L93

I have been trying to create a version where the processes calls `subprocess.Popen(command, env=env_copy, cwd=cwd)` which works as expected but it hangs when calling xm.xla_device(..)

https://github.com/PyTorchLightning/pytorch-lightning/blob/lightning_ddp_tpu/pytorch_lightning/plugins/training_type/tpu_popen.py#L110

I am reusing the code from xmp.spawn to properly set the environment variables and made sure there were the same between TPUSpawn and TPUPopen.

```
    def _setup_xla(self) -> int:
        index = os.getenv("LOCAL_RANK", "0")
        num_processes = os.getenv("PL_NUM_DEVICES")
        dev_kind = os.getenv("PL_DEV_KIND")
        pf_cfg = xmp.PreForkConfig(dev_kind, num_processes)
       # setup all the env variables + no need to setup torch.distributed.init_group.
        xmp._prepare_env_for_index(index, pf_cfg)

        # this hangs
        device = xm.xla_device(index)
        xmp._setup_replication()
        
        return int(index)
```

Would this `TPUPopen` be possible ? And if yes, what I am missing ?

Best,
T.C

**Comment by tchaton** [comment]
@dlibenzi  any thoughts ?

**Comment by zcain117** [comment]
Can you give a summary of why this would be a good thing? Like what kind of performance improvements or usability improvements would we expect?

**Comment by tchaton** [comment]
Hey @zcain117 ,

There would be several reasons. 
Each version (popen vs spawn) has its own advantage. 

A popen version would work with objects which can t be pickled and enable some Pytorch Lightning features to work for TPUs.

It also clean user code as they don t need to wrap their logic within a function to be passed to spawn.

PyTorch 1.9 will actually support both version natively.

Best,
T.C

**Comment by tchaton** [comment]
@JackCaoG

**Comment by rwightman** [comment]
I also prefer this style to spawn and spent a bit of time hacking about to see if I could get it working.

@tchaton I got this working locally on XLA GPU with 2x GPU following a similar path as I. I used a launcher script like the pytorch distributed.launch, called the prefork setup in that, and then at the start of the launched script performed the _prepare_env_for_index and setup_replication...

However, on the TPU it locks up. Seems the client mesh service and xrt services aren't launched properly in that situation, their bring up is missing in the logs and the mesh clients are all stuck trying to connect when the xla device is accessed... eventually times out.

**Comment by rwightman** [comment]
It's not quite clear to me how those lower level service threads (processes?) get kicked off. Any pointers where to look in the code would be appreciated, I'm game to take a look and would likely learn something in the process.

This isn't a high priority for me, but I assumed it would have been a quick fix to get it working.

**Comment by JackCaoG** [comment]
@rwightman For 2vm setup this should not be an issue since server is started in a remote vm.

On the tpuvm non-pod setup, it will be the master process (assuming 8 process multi-processing) to start the local xrt service in the same process. You can find the code in 
https://github.com/pytorch/xla/blob/759e765860bef772b86a80c7c6269249e233b493/third_party/xla_client/xrt_computation_client.cc#L1937
I am suspecting something goes wrong in https://github.com/pytorch/xla/blob/759e765860bef772b86a80c7c6269249e233b493/third_party/xla_client/xrt_computation_client.cc#L172

This is a "hack". Basically it is checking whether the process also carry the `cpu` or `gpu` device. Under the tpuvm multi processing, there should be only process carry the `CPU:0` and `TPU:0`, the rest of the process will carry `TPU:X`. We make that process to start the xrt local service. Adding some debugging messages around here should help us understand why localservice is not being brought up. 

Another quick fix is to start the xrt_local_service in a separate process with 
```
python3 -m torch_xla.core.xrt_run_server --port 51011 --restart
```

Note that master process still needs to get the tpu topo from the local service and start the mesh server to distribute that to clients in this separate server setup.
https://github.com/pytorch/xla/blob/759e765860bef772b86a80c7c6269249e233b493/third_party/xla_client/xrt_computation_client.cc#L1315

**Comment by tchaton** [comment]
Hey @JackCaoG,

Thanks for giving details.

Hey @rwightman, would you like us to collaborate and give it a try ?
If so, please join PyTorch Lightning Slack channel and ping me directly on my GitHub username.

Best,
T.C
